server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java

```diff
@@ -81,16 +91,18 @@ public static void setThreadContext(ThreadContext threadContext) {
      * Remove the {@link ThreadContext} used to add deprecation headers to network responses.
      * <p>
      * This is expected to <em>only</em> be invoked by the {@code Node}'s {@code close} method (therefore once outside of tests).
-     *
+     * Node: This method should be called before closing a <code>ThreadContext</code>.
+     * @see ThreadContext#close()
      * @param threadContext The thread context owned by the {@code ThreadPool} (and implicitly a {@code Node})
-     * @throws IllegalStateException if this {@code threadContext} is unknown (and presumably already unset before)
-     */
-    public static void removeThreadContext(ThreadContext threadContext) {
+     * @return true if the context was removed, false if the context was now known (possibly already removed)
+     **/
+    public static boolean removeThreadContext(ThreadContext threadContext) {
         assert threadContext != null;
-
-        // remove returning false means it did not have it already
```

Comment:
This exception should not be thrown, closing a ThreadPool now also removes  ThreadContext from a DeprecationLogger.
It is possible to close a ThreadPool multiple times as per NodesTests.testAwaitCloseTimeoutsOnNonInterruptibleTask
THis could happen when someone is calling an api multiple times.